poc/vdaf_prio3.sage

```diff
@@ -222,79 +225,90 @@ class Prio3(Vdaf):
+        k_blind = None
         if Prio3.Flp.JOINT_RAND_LEN > 0:
             k_blind, encoded = encoded[:l], encoded[l:]
-            k_hint = []
-            for _ in range(Prio3.SHARES-1):
-                k_joint_rand_part, encoded = encoded[:l], encoded[l:]
-                k_hint.append(k_joint_rand_part)
         if len(encoded) != 0:
             raise ERR_DECODE
-        return (input_share, proof_share, k_blind, k_hint)
+        return (measurement_share, proof_share, k_blind)
 
     @classmethod
     def encode_helper_share(Prio3,
-                            k_input_share,
+                            k_measurement_share,
                             k_proof_share,
-                            k_blind,
-                            k_hint):
+                            k_blind):
         encoded = Bytes()
-        encoded += k_input_share
+        encoded += k_measurement_share
         encoded += k_proof_share
         if Prio3.Flp.JOINT_RAND_LEN > 0:
             encoded += k_blind
-            encoded += concat(k_hint)
         return encoded
 
     @classmethod
     def decode_helper_share(Prio3, dst, agg_id, encoded):
         l = Prio3.Prg.SEED_SIZE
-        k_input_share, encoded = encoded[:l], encoded[l:]
-        input_share = Prio3.Prg.expand_into_vec(Prio3.Flp.Field,
-                                                k_input_share,
-                                                dst + byte(agg_id),
+        k_measurement_share, encoded = encoded[:l], encoded[l:]
+        measurement_share = Prio3.Prg.expand_into_vec(Prio3.Flp.Field,
+                                                k_measurement_share,
+                                                dst + DST_MEASUREMENT_SHARE
+                                                + byte(agg_id),
                                                 Prio3.Flp.INPUT_LEN)
         k_proof_share, encoded = encoded[:l], encoded[l:]
         proof_share = Prio3.Prg.expand_into_vec(Prio3.Flp.Field,
                                                 k_proof_share,
-                                                dst + byte(agg_id),
+                                                dst + DST_PROOF_SHARE 
+                                                + byte(agg_id),
                                                 Prio3.Flp.PROOF_LEN)
-        k_blind, k_hint = None, None
+        k_blind = None
         if Prio3.Flp.JOINT_RAND_LEN > 0:
             k_blind, encoded = encoded[:l], encoded[l:]
-            k_hint = []
-            for _ in range(Prio3.SHARES-1):
+        if len(encoded) != 0:
+            raise ERR_DECODE
+        return (measurement_share, proof_share, k_blind)
+
+    @classmethod
+    def encode_public_share(Prio3,
+                            k_joint_rand_parts):
+        encoded = Bytes()
+        if Prio3.Flp.JOINT_RAND_LEN > 0:
+            encoded += concat(k_joint_rand_parts)
+        return encoded
+
+    @classmethod
+    def decode_public_share(Prio3, encoded):
```

Comment:
The control flow in this function makes it difficult to modify later on. In particular, here you are doing:
```python
thing = X
if special_case:
    consume_buffer
    thing = Y
ensure_buffer_is_empty
return thing
```

The following control flow is preferable, as it makes failures more obvious:
```python
if not special_case:
   ensure_buffer_is_empty
   return X

consume_buffer
ensure_buffer_is_empty
return Y
```